compiler/optimizer/LocalOpts.cpp

```diff
@@ -6932,183 +6928,108 @@ int32_t TR_InvariantArgumentPreexistence::perform()
+               traceIfEnabled("PREX:      No argInfo for parm %d\n", index);
                continue;
                }
 
+            traceIfEnabled("PREX:      Parm %d is arg %p parmInfo %p\n", index, arg, &parmInfo);
+
+            if (arg->hasKnownObjectIndex() && !(arg->getClass() && arg->classIsFixed()))
+               {
+               TR::VMAccessCriticalSection setClass(comp());
+               TR_OpaqueClassBlock *fixedClazz = TR::Compiler->cls.objectClass(comp(), knot->getPointer(arg->getKnownObjectIndex()));
+               arg->setClassIsFixed(fixedClazz);
+               traceIfEnabled("PREX:        Parm %d is known object obj%d\n", index, arg->getKnownObjectIndex());
+               }
+
             bool classIsFixed       = arg->classIsFixed();
             bool classIsPreexistent = arg->classIsPreexistent();
+            if (!classIsFixed && !classIsPreexistent)
+               continue;
 
-            if (classIsFixed || classIsPreexistent)
+            TR_OpaqueClassBlock *clazz = arg->getClass() ? arg->getClass() : clazzFromMethod;
+
+            // bail out if types are not compatible, which is an indicator that we're peeking or inlining dead path
+            if (clazz != clazzFromMethod && fe()->isInstanceOf(clazz, clazzFromMethod, true, true, true) != TR_yes)
+               return 1;
+
+            parmInfo.clear();
+            parmInfo.setSymbol(p);
+            parmInfo.setKnownObjectIndex(arg->getKnownObjectIndex());
+            parmInfo.setClass(clazz);
+            char *sig = NULL;
+            if (trace())
+               sig = TR::Compiler->cls.classSignature(comp(), clazz, trMemory());
+
+            if (clazz != clazzFromMethod)
                {
-               int32_t len = 0;
-               const char   *sig = p->getTypeSignature(len);
-               TR_ASSERT(((*sig == 'L') || (*sig == '[')), "non address argument cannot be fixed/preexistent");
+               parmInfo.setClassIsRefined();
+               traceIfEnabled("PREX:          Parm %d class is refined -- declared as %s\n", index, sig);
+               }
 
-               parmInfo.setSymbol(p);
-               TR_OpaqueClassBlock *clazz = arg->getClass();
-               TR_OpaqueClassBlock *clazzFromMethod = fe()->getClassFromSignature(sig, len, feMethod);
-               TR_ASSERT(!clazz ||
-                         !clazzFromMethod ||
-                         clazz == clazzFromMethod ||
-                         fe()->isInstanceOf(clazz, clazzFromMethod, true, true, true) == TR_yes,
-                         "Type from argInfo should be more specific clazz %p clazzFromMethod %p", clazz, clazzFromMethod);
-
-               if (classIsFixed)
+            if (classIsFixed || TR::Compiler->cls.isClassFinal(comp(), clazz))
+               {
+               parmInfo.setClassIsFixed();
+               traceIfEnabled("PREX:        Parm %d class %p is final %s\n", index, clazz, sig);
+               }
+            else // if (classIsPreexistent)
```

Comment:
this seems a bit confusing can this be made a proper comment or this removed if it is redundant - it seems like commented out code and I'm not sure why it is here or why it was turned off.

Reply:
We early return if the class is not fixed nor is preexistent. I changed this line to `else if (classIsPreexistent) irregardless to avoid confusion.